pkg/upstream/cluster/strict_dns_cluster.go

```diff
@@ -0,0 +1,302 @@
+	defer sdc.mutex.Unlock()
+	sdc.StopResolve()
+
+	// before resolve, init hosts to origin unresolved address
+	sdc.simpleCluster.UpdateHosts(newHosts)
+
+	// update resolve targets and start new resolve task
+	var rts []*ResolveTarget
+	for _, host := range newHosts {
+		addr, port := getHostPortFromAddr(host.AddressString())
+		if addr == "" {
+			if log.DefaultLogger.GetLogLevel() >= log.ERROR {
+				log.DefaultLogger.Errorf("[upstream] [strict_dns_cluster] config address format error: %s", host.AddressString())
+			}
+			continue
+		}
+		// default port: 80
+		if port == "" {
+			port = "80"
+		}
+		config := host.Config()
+		rt := &ResolveTarget{
+			dnsAddress:       addr,
+			port:             port,
+			config:           &config,
+			refreshTimeout:   DefaultRefreshTimeout,
+			dnsRefreshRate:   make(chan time.Duration),
+			stop:             make(chan struct{}),
+			timeout:          make(chan bool),
+			strictDnsCluster: sdc,
+			hosts:            []types.Host{host},
+		}
+
+		utils.GoWithRecover(func() {
+			rt.StartResolve()
+		}, nil)
+		if log.DefaultLogger.GetLogLevel() >= log.DEBUG {
+			log.DefaultLogger.Debugf("[upstream] [strict dns cluster] create a resolver for address: %s", rt.dnsAddress)
+		}
+		rts = append(rts, rt)
+	}
+	sdc.resolveTargets = rts
+}
+
+func (sdc *strictDnsCluster) StopResolve() {
+	for _, rt := range sdc.resolveTargets {
+		rt.StopResolve()
+	}
+}
+
+// calculate next resolve interval
+// 1. if respectDnsTTL configured, and minTtl > 0, use minTtl+1
+// 2. if dnsRefreshRate configured, use dnsRefreshRate
+// 3. use DefaultRefreshInterval(5s)
+func (sdc *strictDnsCluster) calculateNextResolveInterval(minTtl time.Duration) time.Duration {
+	dnsRefreshRate := sdc.dnsRefreshRate
+	if sdc.respectDnsTTL && minTtl > 0 {
+		// increase minTtl by 1 in case we will get dns response with ttl 0
+		dnsRefreshRate = minTtl + time.Second
+	} else if sdc.dnsRefreshRate == 0 {
+		dnsRefreshRate = DefaultRefreshInterval
+	}
+	return dnsRefreshRate
+}
+
+// just check address string, hostname and metadata
+func hostEqual(hosts1, hosts2 *[]types.Host) bool {
+	hosts := map[string]types.Host{}
+	if len(*hosts1) != len(*hosts2) {
+		return false
+	}
+	for _, host := range *hosts1 {
+		hosts[host.AddressString()] = host
+	}
+	for _, host := range *hosts2 {
+		if h, exists := hosts[host.AddressString()]; exists {
+			if h.Hostname() == host.Hostname() && reflect.DeepEqual(h.Metadata(), host.Metadata()) {
+				continue
+			}
+			return false
+		}
+		return false
+	}
+	return true
+}
+
+// combine hosts in current rt and other rt
+func (sdc *strictDnsCluster) updateDynamicHosts(newHosts []types.Host, rt *ResolveTarget) {
+	sdc.mutex.Lock()
+	defer sdc.mutex.Unlock()
+
+	// update current hosts in rt
+	rt.hosts = newHosts
+
+	// append hosts in other resolveTargets to newHosts
+	for _, r := range sdc.resolveTargets {
+		if r.config == rt.config {
+			continue
+		}
+		for _, h := range r.hosts {
+			newHosts = append(newHosts, h)
```

Comment:
May cause the host  increase repeatedly, forexample
domainA: 1.1.1.1
domainB: 1.1.1.2

when domainA resolve response:
domainA：rt.hosts 1.1.1.1
domainB：rt.hosts 1.1.1.2
ALL host:
1.1.1.1
1.1.1.2

when domainB resolve response:
 domainA：rt.hosts 1.1.1.1 、1.1.1.2
 domainB：rt.hosts 1.1.1.2
ALL host:
1.1.1.2
1.1.1.1
1.1.1.2

Reply:
Later function hostSet.setFinalHost implements deduplication. So this place just leave it duplicated.

Reply:
However, the resolution results of the current domain will append the resolution records of other domain.